<a href="https://colab.research.google.com/github/panindarkwah/lab-4-llm-decision-support/blob/main/lab_4_llm_decision_support.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import userdata
from openai import OpenAI

API_KEY = userdata.get("GROQ_API_KEY")

client = OpenAI(
    api_key=API_KEY,
    base_url="https://api.groq.com/openai/v1"
)

MODEL = "llama-3.3-70b-versatile"

print("Client ready.")

Client ready.


In [ ]:
# TODO: Write a helper function you will reuse for the WHOLE lab:
def ask_llm(user_prompt,system_prompt="You are a helpful assistant.",
    temperature=0.7,max_tokens=500):

    response = client.chat.completions.create(
        model=MODEL,
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        temperature=temperature,
        max_tokens=max_tokens,
    )
    return response

# TODO: Call it once with a simple question and print the answer.
response_object = ask_llm("What is computer vision")
print("Response: ")
print(response_object.choices[0].message.content)

#TODO: Print response.usage as well — how many tokens did your call consume?
print("\nUsage:")
# Access the usage from the response object
print(response_object.usage)


Response: 
**Computer Vision: A Comprehensive Overview**

Computer vision is a subfield of artificial intelligence (AI) that enables computers to interpret and understand visual data from the world. It involves the development of algorithms, statistical models, and computer programs that allow computers to process, analyze, and comprehend digital images and videos.

**Key Aspects of Computer Vision:**

1. **Image Processing**: Computer vision involves processing and analyzing digital images to extract relevant information, such as objects, shapes, and patterns.
2. **Object Detection**: Identifying and locating objects within images or videos, including people, animals, vehicles, and other objects.
3. **Image Recognition**: Classifying images into predefined categories, such as recognizing faces, objects, or scenes.
4. **Scene Understanding**: Interpreting the context and meaning of a visual scene, including the relationships between objects and their surroundings.
5. **Machine Learning

**PART 1.1**
1. System role is th set of instructions for the AI. It defines the AI`s functionality, boundaries and how it should respond to a gievn task.
User roles is the actual question or request the user wants the AI to answer.

2. A token is a smallest unit of text that an AI model processes.

3. API providers bill per token because the amount of computation required depends largely on how much text the model processes and generates, not simply on the number of API requests.

In [ ]:
#TODO: Ask the SAME question 5 times at temperature=0.0 and 5 times at temperature=1.2.

question = "Suggest a name for a savings product for market traders in Accra."

# Temperature = 0.0
print("===== Temperature = 0.0 =====")

for i in range(5):
    response = ask_llm(
        question,
        temperature=0.0,
        max_tokens=100
    )
    print(f"\nResponse {i + 1}:")
    print(response.choices[0].message.content)


# Temperature = 1.2
print("\n===== Temperature = 1.2 =====")

for i in range(5):
    response = ask_llm(
        question,
        temperature=1.2,
        max_tokens=100
    )
    print(f"\nResponse {i + 1}:")
    print(response.choices[0].message.content)

# TODO: Print all 10 answers, grouped by temperature.


===== Temperature = 0.0 =====

Response 1:
Here are a few suggestions for a savings product for market traders in Accra:

1. **Makola Save**: "Makola" is a well-known market in Accra, so this name could resonate with market traders.
2. **Trader's Treasure**: This name emphasizes the idea of saving and accumulating wealth.
3. **Accra Amanfu**: "Amanfu" is a Ghanaian word for "savings" or "treasury", so this name incorporates local language

Response 2:
Here are a few suggestions for a savings product for market traders in Accra:

1. **Makola Save**: "Makola" is a well-known market in Accra, so this name could resonate with market traders.
2. **Trader's Treasure**: This name emphasizes the idea of saving and accumulating wealth.
3. **Sika Kokoo**: "Sika" means "money" in the Akan language, and "Kokoo" means "gather" or "collect

Response 3:
Here are a few suggestions for a savings product for market traders in Accra:

1. **Makola Save**: "Makola" is a well-known market in Accra, so this 

**Part 1.2**

With the temperature at 0, the response, where almost the same, just a few one tht gav different response. At emperature 1.2 , the rspone was more varied, al though i sw some mistake in trying to inteprete local language to English.
For a loan system, the low temperature would be more appropriate as, as loan decision should be consistent and predictable  rather than random. Using a low temperature helps ensure the users get similar recommendation and reasoning than inconsistencies in reasoning that woul d crste lots of confusionon the part of the user when ,aking decison on taking loans, due frequent changes in advice.

In [ ]:
LETTERS = {
"L001": """Dear Sir/Madam,
My name is Akosua Mensah and I have been selling provisions at Makola Market for 12 years.
I am applying for a loan of GHS 8,000 to buy a deep freezer and expand into frozen foods.
My current stall makes about GHS 900 profit each month. I have saved GHS 2,500 with your
susu scheme over the past two years and I have never missed a contribution. I can repay
GHS 450 monthly over 20 months. My sister, a teacher, will stand as my guarantor.
Thank you for considering my application.""",

"L002": """Hello,
I am Kwame Boateng, a commercial driver in Kumasi. I need GHS 25,000 urgently to repair my
trotro engine and settle some personal debts. Business has been slow but it will surely
pick up after the festive season. I can pay back whenever the money comes. I do not have
collateral at the moment but God willing everything will be fine. Please help me quickly.""",

"L003": """Dear Loan Committee,
I am Efua Darko, owner of Darko Fashions, a registered dressmaking business in Takoradi
(registration no. BN-2019-4482). I employ three apprentices. I request GHS 15,000 to
purchase two industrial sewing machines and fabric stock ahead of the Christmas season.
Last year my December revenue alone was GHS 22,000; monthly profit averages GHS 2,800.
I hold a fixed deposit of GHS 5,000 with GCB which I can pledge. Proposed repayment:
GHS 1,100 monthly for 15 months. Attached are my sales records for the past 18 months.""",

"L004": """Good day,
My name is Yaw Owusu. I want a loan for my poultry farm at Nsawam. The amount is GHS 12,000
for feed and 500 new layers. I started the farm last year. Sometimes I make good money,
around GHS 1,500 in a good month, but bird flu affected us in March and I lost many birds.
I am rebuilding now. I can repay in 18 months. My uncle has agreed to guarantee the loan
with his taxi.""",

"L005": """Dear Manager,
I am writing on behalf of the Adenta Women's Weaving Cooperative (14 members). We seek
GHS 30,000 to buy a bulk order of yarn directly from the factory, cutting out middlemen and
raising our margins from 15% to about 35%. The cooperative has operated for 6 years and
holds GHS 9,000 in our group account. We propose repayment of GHS 2,000 monthly over
16 months, backed by our group savings and joint liability agreement.""",

"L006": """Hi,
This is Kofi. I saw your advert. I want GHS 50,000 to start a car washing business, a
provision shop, and also import phones from Dubai. I am 22 and full of energy. I have not
started any of these yet but my friends say I am very business minded. I will pay back in
one year when the businesses are booming. No collateral but I am trustworthy.""",
}

# Gold-standard labels for three letters (for Section 4 evaluation):
GOLD = {
  "L001": {"applicant_name": "Akosua Mensah", "amount_ghs": 8000,  "purpose": "buy deep freezer / expand into frozen foods",
           "monthly_profit_ghs": 900,  "has_collateral_or_guarantor": True,  "repayment_months": 20},
  "L003": {"applicant_name": "Efua Darko",    "amount_ghs": 15000, "purpose": "industrial sewing machines and fabric stock",
           "monthly_profit_ghs": 2800, "has_collateral_or_guarantor": True,  "repayment_months": 15},
  "L006": {"applicant_name": "Kofi",          "amount_ghs": 50000, "purpose": "car wash, provision shop, phone imports",
           "monthly_profit_ghs": None, "has_collateral_or_guarantor": False, "repayment_months": 12},
}

print(f"{len(LETTERS)} letters loaded.")


6 letters loaded.


In [ ]:

# TODO: Write SUMMARY_PROMPT_V1 — your first, naive attempt (e.g. just "Summarize this:").
SUMMARY_PROMPT_V1 = "Summarize the important facts in this data: {letter}"

# Run V1 on L002
prompt_L002 = SUMMARY_PROMPT_V1.format(letter=LETTERS["L002"])
response_L002_V1 = ask_llm(prompt_L002, temperature=0.0, max_tokens=200)

print("L002 Summary:")
print(response_L002_V1.choices[0].message.content)


# Run V1 on L006
prompt_L006 = SUMMARY_PROMPT_V1.format(letter=LETTERS["L006"])
response_L006_V1 = ask_llm(prompt_L006, temperature=0.0, max_tokens=200)

print("\nL006 Summary:")
print(response_L006_V1.choices[0].message.content)

#TODO: Now write SUMMARY_PROMPT_V2 as a proper template with:
SUMMARY_PROMPT_V2_SYSTEM = """
You are an assistant to a microfinance loan officer.

Summarize loan applications into a brief, factual, and neutral summary.
Use only information provided in the application.
Do not invent, assume, or infer any details.
Keep the summary to 3-4 sentences and focus on information relevant to the loan officer.
"""

def create_summary_prompt_v2(letter_text):
    return f"Summarize this loan application:\n\n{letter_text}"


#  Run V2 on L002
prompt = create_summary_prompt_v2(LETTERS['L002'])

response_L002_V2 = ask_llm(
    prompt,
    system_prompt=SUMMARY_PROMPT_V2_SYSTEM,
    temperature=0.0,
    max_tokens=200
)

print("======== L002 V2 ==========")
print(response_L002_V2.choices[0].message.content)


# Run V2 on L006
prompt = create_summary_prompt_v2(LETTERS['L006'])

response_L006_V2 = ask_llm(
    prompt,
    system_prompt=SUMMARY_PROMPT_V2_SYSTEM,
    temperature=0.0,
    max_tokens=200
)

print("\n===== L006 V2 =========")
print(response_L006_V2.choices[0].message.content)

# Compare V1 vs V2 for L002 and L006

print("=" * 80)
print("L002 — V1 vs V2")
print("=" * 80)

print("\n--- V1 ---")
print(response_L002_V1.choices[0].message.content)

print("\n--- V2 ---")
print(response_L002_V2.choices[0].message.content)


print("\n" + "=" * 80)
print("L006 — V1 vs V2")
print("=" * 80)

print("\n--- V1 ---")
print(response_L006_V1.choices[0].message.content)

print("\n--- V2 ---")
print(response_L006_V2.choices[0].message.content)

L002 Summary:
Here are the important facts from the data:

* The person's name is Kwame Boateng.
* He is a commercial driver in Kumasi.
* He needs GHS 25,000 (Ghanaian Cedi) urgently.
* The money is needed to:
	+ Repair his trotro (minibus) engine.
	+ Settle some personal debts.
* He expects his business to improve after the festive season.
* He has no collateral to offer at the moment.
* He is asking for help and promises to pay back the money when he can.

L006 Summary:
Here are the important facts from the data:

* The applicant, Kofi, is 22 years old.
* He is requesting GHS 50,000 to start three businesses: a car washing business, a provision shop, and a phone import business from Dubai.
* He has not yet started any of these businesses.
* He claims to be "business-minded" based on feedback from his friends.
* He promises to repay the loan within one year, expecting the businesses to be successful by then.
* He does not have any collateral to offer, but assures that he is trustworth

**Part 3.1**

1. V1 just went ahead and wrote a lot of details while V2 was specifically instructed to produce 3 -4 sentences, so produced concise relevant information in 4 sentences. V2 included unneccessary information. For exampl adding approximately $ 8,500 USD, depending on the exchange rate. The dollar application was requesting GHS 50,000, so the dollar conversion was unnecessary for the requested summary. V2 removed it.

2. No invented details” is essential because this is a loan application, which can affect a real financial decision. If the LLM adds information that the applicant never provided, a loan officer could make an incorrect assessment of the applicant.
This failure is called hallucination.

In [ ]:
# TODO: Write EXTRACT_PROMPT — a template that instructs the model to return ONLY a JSON
#   object with EXACTLY these keys:
#     applicant_name (string), amount_ghs (number), purpose (string),
#     monthly_profit_ghs (number or null), has_collateral_or_guarantor (boolean),
#     repayment_months (number or null)

EXTRACT_PROMPT = """
You are an assistant that extracts structured information from loan applications.

Return ONLY a valid JSON object with EXACTLY these six keys:

{{
  "applicant_name": "string",
  "amount_ghs": 0,
  "purpose": "string",
  "monthly_profit_ghs": null,
  "has_collateral_or_guarantor": false,
  "repayment_months": null
}}

Rules:
- Return ONLY the JSON object. Do not include explanations, comments, or markdown.
- Use EXACTLY the six keys specified above.
- applicant_name must be a string.
- amount_ghs must be a number.
- purpose must be a string.
- monthly_profit_ghs must be a number if explicitly stated; otherwise use null.
- has_collateral_or_guarantor must be true if the applicant explicitly has collateral or a guarantor, and false if the applicant explicitly states that they do not.
- repayment_months must be a number if the repayment period is explicitly stated; otherwise use null.
- If a field is not stated in the letter, use null. Do not guess.
- Do not infer information that is not explicitly provided.

Worked example:

Letter:
"My name is Ama Mensah. I run a small bakery in Accra and I am requesting GHS 15,000 to purchase a new oven and raw materials. My business earns a monthly profit of GHS 2,000. My brother will act as my guarantor, and I plan to repay the loan within 10 months."

Correct JSON:
{{
  "applicant_name": "Ama Mensah",
  "amount_ghs": 15000,
  "purpose": "Purchase a new oven and raw materials",
  "monthly_profit_ghs": 2000,
  "has_collateral_or_guarantor": true,
  "repayment_months": 10
}}

Now extract the information from this loan application:

{letter}
"""


letter_acheampong = """
My name is Acheampong. I am a student, currently studying computer science.
I am currently building a project to help farmers keep track of their livestock,
crop yields and the gains they get after every harvest. I need GHS 10,000 to
fund my research activities and hire the services of a software engineer.
I believe I would be able to pay it back in two years. I do not have a
guarantee, but I believe my project will be completed within months, and I
will be able to get my returns soon, which will help me in the payment of
the loan.
"""

prompt = EXTRACT_PROMPT.format(letter=letter_acheampong)

response = ask_llm(
    prompt,
    temperature=0,
    max_tokens=300
)

print(response.choices[0].message.content)


import json

def extract_fields(letter_text):
    prompt = EXTRACT_PROMPT.format(letter=letter_text)

    response = ask_llm(
        prompt,
        temperature=0,
        max_tokens=300
    )

    result = response.choices[0].message.content.strip()


# Remove ```json and ``` fences if the model includes them
    if result.startswith("```json"):
        result = result[len("```json"):].strip()

    if result.startswith("```"):
        result = result[len("```"):].strip()

    if result.endswith("```"):
        result = result[:-3].strip()

    # Try to parse the JSON
    try:
        return json.loads(result)

    except json.JSONDecodeError as e:
        print(f"Warning: Failed to parse LLM response as JSON: {e}")
        return None

result = extract_fields(letter_acheampong)
print(result)


import pandas as pd

# Extract fields from all six letters
results = []

for letter_id, letter_text in LETTERS.items():
    extracted = extract_fields(letter_text)

    if extracted is not None:
        extracted["letter_id"] = letter_id
        results.append(extracted)

# Create DataFrame
df = pd.DataFrame(results)

# Put letter_id first
columns = ["letter_id", "applicant_name", "amount_ghs", "purpose",
           "monthly_profit_ghs", "has_collateral_or_guarantor",
           "repayment_months"]

df = df[columns]

# Display the DataFrame
display(df)






{
  "applicant_name": "Acheampong",
  "amount_ghs": 10000,
  "purpose": "Fund research activities and hire the services of a software engineer",
  "monthly_profit_ghs": null,
  "has_collateral_or_guarantor": false,
  "repayment_months": 24
}
{'applicant_name': 'Acheampong', 'amount_ghs': 10000, 'purpose': 'Fund research activities and hire the services of a software engineer', 'monthly_profit_ghs': None, 'has_collateral_or_guarantor': False, 'repayment_months': 24}


,letter_id,applicant_name,amount_ghs,purpose,monthly_profit_ghs,has_collateral_or_guarantor,repayment_months
0,L001,Akosua Mensah,8000,Buy a deep freezer and expand into frozen foods,900.0,True,20.0
1,L002,Kwame Boateng,25000,Repair trotro engine and settle personal debts,NaN,False,NaN
2,L003,Efua Darko,15000,Purchase two industrial sewing machines and fa...,2800.0,True,15.0
3,L004,Yaw Owusu,12000,For feed and 500 new layers,1500.0,True,18.0
4,L005,Adenta Women's Weaving Cooperative,30000,Buy a bulk order of yarn directly from the fac...,NaN,True,16.0
5,L006,Kofi,50000,"Start a car washing business, a provision shop...",NaN,False,12.0


**PART 3.2**
1. The few-shot example should not come from the six letters being processed because it would cause the modl to memorize the information from the example istead of actually extracting infromation from the new letter.  Using a cmpletely different example, made it possible to test whether the model actually learned the extraction rules and JSON format, rather than simply copying facts from one of the the letters.

2. The "use null, do not guess" instuction is important, because without it the model may try to fill inmissing information using assumptions or prediction based on information it knows fromother datasets.

3. The temperature 0 is appropriate for the structured extraction because the goal is to get a consistent  and fact-based output.

In [ ]:
BRIEF_PROMPT = """
You are an assistant to a microfinance loan officer.

Review the loan application and the extracted information below.
Provide a brief assessment to help the loan officer decide what
information or follow-up may be needed.

IMPORTANT:
- Use only information provided in the loan application and extracted JSON.
- Do not invent, assume, or infer facts.
- Identify every strength and risk in the information provided.
- If information is missing, identify it as missing rather than guessing.
- The final loan decision must always be made by a human loan officer.
- Do NOT recommend approving or rejecting the loan.

Your response MUST contain exactly these four sections:

1. Strengths
- Provide concise bullet points about positive or relevant factors
  explicitly supported by the application.

2. Risks / Red Flags
- Provide concise bullet points about potential concerns or risks
  explicitly supported by the application.

3. Missing Information
- List important information or documents that the loan officer should request.

4. Suggested Next Step
- Recommend an appropriate follow-up action, such as:
  "invite for interview", "request documents", or "flag for senior review".
- Do not say "approve" or "reject".

Remember: You are providing decision-support only. The final lending
decision must be made by a human.

Loan Application:
{letter}

Extracted Information:
{extracted_json}
"""


# TODO: Generate briefs for ALL SIX letters. Print the briefs for L001, L002, and L006 —
#   three very different applications.
import json

briefs = {}

# Generate a brief for all six letters
for letter_id, letter_text in LETTERS.items():

    # Extract structured information
    extracted = extract_fields(letter_text)

    if extracted is None:
        print(f"Warning: Could not extract fields for {letter_id}")
        continue

    # Create the brief prompt
    prompt = BRIEF_PROMPT.format(
        letter=letter_text,
        extracted_json=json.dumps(extracted, indent=2)
    )

    # Call the LLM
    response = ask_llm(
        prompt,
        temperature=0.0,
        max_tokens=500
    )

    # Store the brief
    briefs[letter_id] = response.choices[0].message.content


# Print briefs for L001, L002, and L006
for letter_id in ["L001", "L002", "L003", "L006"]:
    print("=" * 80)
    print(f"{letter_id} — LOAN OFFICER BRIEF")
    print("=" * 80)
    print(briefs[letter_id])
    print()


L001 — LOAN OFFICER BRIEF
## 1. Strengths
* The applicant, Akosua Mensah, has 12 years of experience selling provisions at Makola Market.
* She has a stable monthly profit of GHS 900 from her current stall.
* She has saved GHS 2,500 with the susu scheme over two years without missing a contribution, demonstrating a good savings habit.
* She has a guarantor, her sister, who is a teacher, adding a level of security to the loan.
* She has a clear plan for repayment, proposing to pay GHS 450 monthly over 20 months.

## 2. Risks / Red Flags
* The loan amount of GHS 8,000 is significant compared to her monthly profit, which might pose a risk if her business does not expand as planned.
* There is no detailed information on how the expansion into frozen foods will affect her current business operations or profitability.
* The repayment plan relies on her ability to maintain or increase her current monthly profit, which may be subject to market fluctuations.

## 3. Missing Information
* Detaile

**PART 3.3**

1. The system was able to identify the right right strengths and red felgs in each. In  L003, the model was able to identify, the right funds the Efua, neded, the monthly profit she makes and her collatoral,the plans she had for the repayment of the loa if granted one. This is exactly the needed factors that a loan oofcer needs to decide whether to grant access or not.

 The system identfied very relevant red flags, that woukd make the applicant unable to pay ba k the loan. The system suggested that the proopesed GHS 1,100 monthly repayment for 15 months could be difficult if profits fluctuate.This a very good red flag spotted as, business profits are not alwsys fixed so there is no quarantee that, she is sure gooing to make that profit every month.

 The system idetfied very relevant strengths in L006.  The system identified Kofi has clear purpose for the loan, thats car wash, provison, provio shop and phone import business.
 With the red flags the system identified, that none of the businesses has statrted yet, has no collatoral or quarantor  and the payment is dependent on when the Business becomes successful. All these are valid factors that would hinder its chances of getting the loan.

 2. The model can make mistakes or misunderstand information in the application if it directly says "approve" or " reject",  a loan officer would rely on that, without having to review the request by him or herself. loans have high impact on bith the lifes of the appilcant and the operatio of the firm granting the loa, so rely on a n AI system to make final decison could lead unfair or baised outomes, and mostly affect the reputation of the firm and in worst cases cost finacial bruden to the firm, if eleigible applicant are gieven approval and unable to pay back.



**PART 3.4**

 738d5ee3f0602a93785010a38b5f3cecfa6fb952

In [ ]:
# TODO: For the three letters in GOLD, compare your extracted DataFrame to the gold values
#   field by field. Compute per-field accuracy across the three letters
#   (name matching can be case-insensitive; numbers must match exactly).

import pandas as pd

# Fields to evaluate
fields = [
    "applicant_name",
    "amount_ghs",
    "purpose",
    "monthly_profit_ghs",
    "has_collateral_or_guarantor",
    "repayment_months"
]

# The three letters in GOLD
gold_ids = list(GOLD.keys())

results = []

for field in fields:
    correct = 0

    for letter_id in gold_ids:
        # Get extracted value
        extracted_value = df.loc[
            df["letter_id"] == letter_id, field
        ].iloc[0]

        # Get gold value
        gold_value = GOLD[letter_id][field]

        # Name comparison: case-insensitive
        if field == "applicant_name":
            match = (
                str(extracted_value).strip().lower()
                == str(gold_value).strip().lower()
            )
        else:
            # Numbers and other fields must match exactly
            match = extracted_value == gold_value

        if match:
            correct += 1

    accuracy = correct / len(gold_ids)

    results.append({
        "field": field,
        "correct": correct,
        "total": len(gold_ids),
        "accuracy": accuracy
    })

accuracy_df = pd.DataFrame(results)

# Display accuracy as percentages
accuracy_df["accuracy"] = accuracy_df["accuracy"] * 100

display(accuracy_df)


# TODO: Display a small table: rows = fields, columns = L001 / L003 / L006 / accuracy.
fields = [
    "applicant_name",
    "amount_ghs",
    "purpose",
    "monthly_profit_ghs",
    "has_collateral_or_guarantor",
    "repayment_months"
]

gold_ids = ["L001", "L003", "L006"]

results = []

for field in fields:
    row = {"field": field}
    correct_count = 0

    for letter_id in gold_ids:
        extracted_value = df.loc[
            df["letter_id"] == letter_id, field
        ].iloc[0]

        gold_value = GOLD[letter_id][field]

        # Case-insensitive comparison for names
        if field == "applicant_name":
            match = (
                str(extracted_value).strip().lower()
                == str(gold_value).strip().lower()
            )
        else:
            match = extracted_value == gold_value

        row[letter_id] = "✓" if match else "✗"

        if match:
            correct_count += 1

    row["accuracy"] = f"{correct_count / 3 * 100:.1f}%"
    results.append(row)

comparison_df = pd.DataFrame(results)

display(comparison_df)

,field,correct,total,accuracy
0,applicant_name,3,3,100.000000
1,amount_ghs,3,3,100.000000
2,purpose,0,3,0.000000
3,monthly_profit_ghs,2,3,66.666667
4,has_collateral_or_guarantor,3,3,100.000000
5,repayment_months,3,3,100.000000


,field,L001,L003,L006,accuracy
0,applicant_name,✓,✓,✓,100.0%
1,amount_ghs,✓,✓,✓,100.0%
2,purpose,✗,✗,✗,0.0%
3,monthly_profit_ghs,✓,✓,✗,66.7%
4,has_collateral_or_guarantor,✓,✓,✓,100.0%
5,repayment_months,✓,✓,✓,100.0%


In [ ]:
import json

# Part 4.1
# TODO: Run extract_fields() on letter L004 FIVE times at temperature=0 and FIVE times at
#   temperature=1.0.

def extract_fields(letter_text, temperature=0):
    prompt = EXTRACT_PROMPT.format(letter=letter_text)

    response = ask_llm(
        prompt,
        temperature=temperature,
        max_tokens=300
    )

    result = response.choices[0].message.content.strip()

    # Remove ```json fences
    if result.startswith("```json"):
        result = result[len("```json"):].strip()

    if result.startswith("```"):
        result = result[len("```"):].strip()

    if result.endswith("```"):
        result = result[:-3].strip()

    try:
        return json.loads(result)

    except json.JSONDecodeError as e:
        print(f"Warning: Failed to parse JSON: {e}")
        return None


#TODO: For each temperature, report how many of the 5 runs produced (a) valid JSON and
#   (b) identical values across runs. A simple approach: json.dumps(result, sort_keys=True)
#   and count unique strings.

def evaluate_runs(results):
    # Count valid JSON results
    valid_results = [r for r in results if r is not None]

    # Convert dictionaries to consistent strings
    json_strings = [
        json.dumps(r, sort_keys=True)
        for r in valid_results
    ]

    # Count unique outputs
    unique_outputs = len(set(json_strings))

    return len(valid_results), unique_outputs


# --- Run extract_fields for L004 five times at temperature=0 ---
results_temp_0 = []
for _ in range(5):
    results_temp_0.append(extract_fields(LETTERS["L004"], temperature=0))

# --- Run extract_fields for L004 five times at temperature=1.0 ---
results_temp_1 = []
for _ in range(5):
    results_temp_1.append(extract_fields(LETTERS["L004"], temperature=1.0))


# Evaluate temperature = 0
valid_0, unique_0 = evaluate_runs(results_temp_0)

# Evaluate temperature = 1.0
valid_1, unique_1 = evaluate_runs(results_temp_1)


print("Temperature = 0")
print(f"Valid JSON: {valid_0}/5")
print(f"Unique outputs: {unique_0}")
print(f"Identical across all runs: {'Yes' if unique_0 == 1 else 'No'}")

print("\nTemperature = 1.0")
print(f"Valid JSON: {valid_1}/5")
print(f"Unique outputs: {unique_1}")
print(f"Identical across all runs: {'Yes' if unique_1 == 1 else 'No'}")

Temperature = 0
Valid JSON: 5/5
Unique outputs: 1
Identical across all runs: Yes

Temperature = 1.0
Valid JSON: 5/5
Unique outputs: 2
Identical across all runs: No


In [ ]:
#Test 2 — Feed your extractor an EMPTY or IRRELEVANT text (e.g. a weather report).
#     Does it return nulls, or does it fabricate an applicant?

# Adversarial Test 1
# Ask for information that is not provided in the letter.

ADVERSARIAL_PROMPT_1 = """
What is the applicant's credit score?
If the credit score is not stated in the letter, say that it is not provided.
Do not guess or invent a value.

Loan application:
{letter}
"""

prompt = ADVERSARIAL_PROMPT_1.format(letter=LETTERS["L002"])

response = ask_llm(
    prompt,
    system_prompt=SUMMARY_PROMPT_V2_SYSTEM,
    temperature=0,
    max_tokens=100
)

print(response.choices[0].message.content)

#Test 2 — Feed your extractor an EMPTY or IRRELEVANT text (e.g. a weather report).
#Does it return nulls, or does it fabricate an applicant?

# Adversarial Test 2
# Give the extractor text that contains no loan application information.

irrelevant_text = """
Today is a cloudy Saturday morning, plans of going to the gym seems like it would not come to reality.
I don`t want to be beaten by the rain. I am in  a dilemma right now , whether to go or not.Checking from
the fact that I have missed chest day and shoulder this week, I can`t afford to miss today to. Help me, to decide
will there be rains today. """




result = extract_fields(irrelevant_text)

print("===== Adversarial Test 2 =====")
print(result)

# Test 1
# The applicant, Kwame Boateng, is requesting a loan of GHS 25,000 to repair his trotro engine and settle personal debts. He mentions that business has been slow, but expects it to improve after the festive season. The applicant does not have collateral to offer at the moment. The credit score is not provided in the loan application.
#Result : PASS

# {'applicant_name': None, 'amount_ghs': None, 'purpose': None, 'monthly_profit_ghs': None, 'has_collateral_or_guarantor': False, 'repayment_months': None}
# Result: False.







The applicant, Kwame Boateng, is requesting a loan of GHS 25,000 to repair his trotro engine and settle personal debts. He mentions that business has been slow, but expects it to improve after the festive season. The applicant does not have collateral to offer at the moment. The credit score is not provided in the loan application.
===== Adversarial Test 2 =====
{'applicant_name': None, 'amount_ghs': None, 'purpose': None, 'monthly_profit_ghs': None, 'has_collateral_or_guarantor': False, 'repayment_months': None}


**PART 4**

The extraction accuracy was 100% for applicant_name, amount_ghs, has_collateral_or_guarantor, and repayment_months. monthly_profit_ghs achieved 66.7% accuracy, while purpose was the hardest field, with only 0% accuracy across the three letters.

The purpose field was hardest because loan purposes are expressed in natural language and can contain multiple activities. The model may extract the correct meaning but phrase it differently from the exact wording in the GOLD values. Since the evaluation requires an exact match for this field, semantically similar purposes are still counted as incorrect.

2. The reliability experiment showed that temperature 0 produced completely consistent results: all 5 runs returned valid JSON and all 5 outputs were identical. At temperature 1.0, all 5 runs still produced valid JSON, but there were 2 unique outputs, meaning the results were not completely consistent across runs. This shows that higher temperature can introduce variability, even when the output format remains valid. Therefore, for production systems that require reliable and predictable structured extraction, temperature 0 is preferable.

3. The system did not hallucinate in Test 1. When asked for the applicant`s credit score, it correctly stated that the credit score was not provided in the application. However, it di show a failure in test2, when given an irrelevant weather report, the extraxtor retruned false for has collatoral or quarantor instead of none. This form of unsupported inference because the weather report contained no information about an applicant or collateral. To reduce this risk, the prompt should explicitly sate that if the input does not contain a loan application or the required information.



**PART 4.4 **
1. The system should not be used to fully automate apporval or rejection. If the bank relied entirely on the system, applicant who are not too conversant with English or formal way of writing face a challenge of failing out of being granted access to loans despite havinh well established business. Applicants with poor wrting skills would be harm despite having good business running, neccessarly credentail for them to access a loan. The system should ther be used to extract relevant informatiins to the loam user, while the final decison rely on humans.

2. Sending loan application to a third-party API in another country means that personal and financial information is being transferered outside Ghana and processed by an external organization. Before deploying this system at a Ghanaian microfinance institution, I would check complaince with Ghana's Data Protection Act, 2012 (Act 843), including whether the institution has a lawful basis for processing the data and whether appropriate consent, privacy notices, security safeguards, and international data-transfer arrangements are in place. I would also review the API provider's data-processing terms, where the data is stored, who can access it, whether customer data is retained or used to train models, breach-notification procedures, encryption and deletion policies.

3. The two concrete safeguards I will build are;
mandatory AI reviews  and continous monitory and audit logging.
The AI would never make final decsion of accepting or ejecting a loan. The final decsion would always be reviewd by a loan offocer based on the summary of information Gotten from the AI. Secondly, frequent audits will be conducted to examine inputs, extracted fields, identified fields and human decisions. This will help detect hallucinations and any other errors.


**Section 5**
1. Prompt iteration is similar to hyperparameter tuning because both involve systematically changing inputs to a system, running experiments, and evaluating the resulting performance rather than relying on intuition. In Lab 3, I changed hyperparameters such as learning rate, dropout, and network architecture and compared their effects on model performance. In this lab, I changed the prompt structure, adding role instructions, output constraints, schemas, examples, and rules against guessing, and evaluated the effects using extraction accuracy, consistency, and adversarial tests. The key difference is that hyperparameter tuning changes how a model is configured or trained, while prompt engineering changes the instructions given to an already-trained model without changing its learned parameters.

2. Yes, I would trust the system to run unattended for the specific task of extracting and organizing loan information, because the evaluation showed strong performance on most fields. The extraction achieved 100% accuracy for applicant name, loan amount, collateral/guarantor status, and repayment period, while monthly profit achieved 66.7% accuracy. The reliability test also showed that at temperature 0, all five runs produced valid JSON and exactly the same output, demonstrating strong consistency. Although the adversarial test revealed a weakness when irrelevant text was provided, this could be addressed by strengthening the null-handling rule and validating the input before extraction. Therefore, I would trust it for automated extraction and structuring, while still keeping final loan decisions separate from the system.

3. The API used 545 tokens in total for the request: 45 prompt tokens and 500 completion tokens. Assuming approximately 545 tokens per application, processing 1,000 applications per month would require about 545,000 tokens per month. This suggests that I should compare providers based not only on token pricing but also on privacy, security, reliability, and data-retention policies because the system processes sensitive loan applications. A cheaper provider may not necessarily be the best choice if it has weaker privacy protections.

4.  For this task, using a foundation model through an API is preferable because the model already has strong language-understanding capabilities, allowing us to summarize applications, extract structured information, and identify risks with relatively little development and no model training. This is faster and requires less data and computational resources than training our own neural network. However, training our own model could be preferable when we have a large, high-quality domain-specific dataset, need highly specialized behavior, require complete control over the model and data, or need to process sensitive information without sending it to a third-party provider.